scripts/esrupgrades/esr.5.37-7.8.mysql.up.sql

```diff
@@ -1,1295 +1,1376 @@
+		GROUP BY index_name
+		INTO AlterIndex;
+
+	SELECT COUNT(*) FROM INFORMATION_SCHEMA.COLUMNS
+		WHERE table_name = 'UploadSessions'
+		AND table_schema = DATABASE()
+		AND column_name = 'Type'
+		AND REPLACE(LOWER(column_type), '"', "'") != "enum('attachment','import')"
+		INTO AlterColumn;
+
+	IF CreateIndex THEN
+		SET CreateIndexQuery = 'ADD INDEX idx_uploadsessions_type (Type)';
+	END IF;
+
+	IF AlterIndex THEN
+		SET AlterIndexQuery = 'DROP INDEX idx_uploadsessions_user_id, ADD INDEX idx_uploadsessions_user_id (UserId)';
+	END IF;
+
+	IF AlterColumn THEN
+		SET AlterColumnQuery = 'MODIFY COLUMN Type ENUM("attachment", "import")';
+	END IF;
+
+	IF CreateIndex OR AlterIndex OR AlterColumn THEN
+		SET @query = CONCAT('ALTER TABLE UploadSessions ', CONCAT_WS(', ', CreateIndexQuery, AlterIndexQuery, AlterColumnQuery));
+		PREPARE stmt FROM @query;
+		EXECUTE stmt;
+		DEALLOCATE PREPARE stmt;
 	END IF;
 END//
 DELIMITER ;
-CALL MigrateCRTChannelMembershipCounts ();
-DROP PROCEDURE IF EXISTS MigrateCRTChannelMembershipCounts;
+SELECT CONCAT('-- ', NOW(), ' MigrateUploadSessions procedure starting.') AS DEBUG;
+CALL MigrateUploadSessions ();
+SELECT CONCAT('-- ', NOW(), ' MigrateUploadSessions procedure finished.') AS DEBUG;
+DROP PROCEDURE IF EXISTS MigrateUploadSessions;
 
 /* ==> mysql/000055_create_crt_thread_count_and_unreads.up.sql <== */
 /* fixCRTThreadCountsAndUnreads Marks threads as read for users where the last
 reply time of the thread is earlier than the time the user viewed the channel.
 Marking a thread means setting the mention count to zero and setting the
 last viewed at time of the the thread as the last viewed at time
 of the channel */
-
 DELIMITER //
-CREATE PROCEDURE MigrateCRTThreadCountsAndUnreads ()
+CREATE PROCEDURE MigrateThreadMemberships ()
 BEGIN
-	IF(SELECT EXISTS(SELECT * FROM Systems WHERE Name = 'CRTThreadCountsAndUnreadsMigrationComplete') = 0) THEN
-		UPDATE
-			ThreadMemberships
-			INNER JOIN (
-				SELECT
-					PostId,
-					UserId,
-					ChannelMembers.LastViewedAt AS CM_LastViewedAt,
-					Threads.LastReplyAt
-				FROM
-					Threads
-					INNER JOIN ChannelMembers ON ChannelMembers.ChannelId = Threads.ChannelId
-				WHERE
-					Threads.LastReplyAt <= ChannelMembers.LastViewedAt) AS q ON ThreadMemberships.Postid = q.PostId
-				AND ThreadMemberships.UserId = q.UserId SET LastViewed = q.CM_LastViewedAt + 1, UnreadMentions = 0, LastUpdated = (
-				SELECT
-					(SELECT ROUND(UNIX_TIMESTAMP(NOW(3))*1000)));
-		INSERT INTO Systems
-			VALUES('CRTThreadCountsAndUnreadsMigrationComplete', 'true');
+	-- UPDATE ThreadMemberships SET LastViewed = ..., UnreadMentions = ..., LastUpdated = ...
+	DECLARE UpdateThreadMemberships BOOLEAN;
+	DECLARE UpdateThreadMembershipsQuery TEXT DEFAULT NULL;
+
+	SELECT COUNT(*) = 0 FROM Systems
+		WHERE Name = 'CRTThreadCountsAndUnreadsMigrationComplete'
+		INTO UpdateThreadMemberships;
+
+	IF UpdateThreadMemberships THEN
+		UPDATE ThreadMemberships INNER JOIN (
```

Comment:
We could, although if they happen, they'd already be logged in the query log they should be using when running the script, right? I'm 0/5, if you think it can help, we could add some more statements here.

Reply:
I'm actually unsure about the query log, especially given it's inside a stored procedure? Hopefully, yes! I'm 0/5 here too.